File: src/main/java/eu/xenit/care4alf/ContentStore.java
package eu.xenit.care4alf;

import com.github.dynamicextensionsalfresco.annotations.AlfrescoService;
import com.github.dynamicextensionsalfresco.annotations.ServiceType;
import com.github.dynamicextensionsalfresco.webscripts.annotations.*;
import com.github.dynamicextensionsalfresco.webscripts.resolutions.JsonWriterResolution;
import com.github.dynamicextensionsalfresco.webscripts.resolutions.Resolution;
import eu.xenit.care4alf.helpers.UtilHelper;
import org.alfresco.model.ContentModel;
import org.alfresco.query.PagingRequest;
import org.alfresco.repo.domain.node.ContentDataWithId;
import org.alfresco.repo.security.authentication.AuthenticationUtil;
import org.alfresco.service.ServiceRegistry;
import org.alfresco.service.cmr.repository.ContentService;
import org.alfresco.service.cmr.repository.NodeRef;
import org.alfresco.service.cmr.repository.NodeService;
import org.alfresco.service.cmr.security.PersonService;
import org.alfresco.service.namespace.QName;
import org.json.JSONException;
import org.json.JSONWriter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.extensions.webscripts.WebScriptResponse;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Component;

import javax.sql.DataSource;
import java.util.*;

/**
 * Created by Thomas.Straetmans on 10/05/2017.
 */

@Component
@WebScript(baseUri = "/xenit/care4alf/contentstore", families = "care4alf", description = "Content store verification")
@Authentication(AuthenticationType.ADMIN)
public class ContentStore {

    @Autowired
    ServiceRegistry services;

    @Autowired
    @AlfrescoService(ServiceType.LOW_LEVEL)
    private ContentService contentService;

    @Autowired
    @AlfrescoService(ServiceType.LOW_LEVEL)
    private NodeService nodeService;

    final private Logger logger = LoggerFactory.getLogger(ContentStore.class);

    private JdbcTemplate jdbcTemplate;// = new JdbcTemplate(dataSource);

    private QName WorkspaceDiskUsage = QName.createQName("WorkspaceDiskUsage");
    private QName ArchiveDiskUsage = QName.createQName("ArchiveDiskUsage");

    @Autowired
    public ContentStore(DataSource dataSource) {
        jdbcTemplate = new JdbcTemplate(dataSource);
    }

    @Uri(value = "/diskusagebyowner")
    @Transaction(readOnly = true)
    public Resolution list(final WebScriptResponse response) {
        List<PersonService.PersonInfo> people = services.getPersonService().getPeople(null, null, null, new PagingRequest(Integer.MAX_VALUE, null)).getPage();
        List<NodeRef> peopleref = new ArrayList<>(people.size());
        logger.error("# of pll found: {}", people.size());
        for (PersonService.PersonInfo info : people) {
            peopleref.add(info.getNodeRef());
        }
        final ArrayList<Map<String, Object>> responselist = new ArrayList();
        for (NodeRef ref : peopleref) {

            Long workspace = (Long) services.getNodeService().getProperty(ref, WorkspaceDiskUsage);
            Long archive = (Long) services.getNodeService().getProperty(ref, ArchiveDiskUsage);
            logger.error("For {} we have found archive: {}, workspace: {}", ref, archive, workspace);
            if (workspace != null || archive != null) {
                Map<String, Object> responsemap = new HashMap<>();
                responsemap.put("username", services.getNodeService().getProperty(ref, ContentModel.PROP_USERNAME));
                responsemap.put("workspace", workspace);
                responsemap.put("archive", archive);
                responselist.add(responsemap);
            }
        }

        return new JsonWriterResolution() {
            @Override
            protected void writeJson(JSONWriter jsonWriter) throws JSONException {
                jsonWriter.array();
                for (Map<String, Object> map : responselist) {
                    jsonWriter.object();
                    for (Map.Entry<String, Object> entry : map.entrySet()) {
                        jsonWriter.key(entry.getKey()).value(entry.getValue());
                    }
                    jsonWriter.endObject();
                }
                jsonWriter.endArray();
            }
        };
    }

    @Uri(value = "/updatediskusage", method = HttpMethod.PUT)
    public void update() {
        AuthenticationUtil.runAs(new AuthenticationUtil.RunAsWork<Object>() {

            @Override
            public Void doWork() throws Exception {
                logger.error("Update diskusage started");
                Calendar timestart = Calendar.getInstance();
                HashMap<String, Long> workspace = new HashMap<>();
                HashMap<String, Long> archive = new HashMap<>();
                List<Long> ids = jdbcTemplate.queryForList("select id from alf_node", Long.class);
                for (Long id : ids) {
                    NodeRef ref = services.getNodeService().getNodeRef(id);
                    if (ref != null) {
                        ContentDataWithId data = (ContentDataWithId) services.getNodeService().getProperty(ref, ContentModel.PROP_CONTENT);
                        Object ownerobj = services.getNodeService().getProperty(ref, ContentModel.PROP_CREATOR);
                        String owner = (ownerobj == null ? null : ownerobj.toString());
                        if (data != null && owner != null) {
                            if (ref.getStoreRef().getProtocol().equals("archive")) {
                                archive.put(owner, (Long) UtilHelper.getOrElse(archive, owner, 0L) + data.getSize());
                            } else {
                                workspace.put(owner, (Long) UtilHelper.getOrElse(workspace, owner, 0L) + data.getSize());
                            }
                        }
                    }
                }

                for (Map.Entry<String, Long> entry : workspace.entrySet()) {
                    NodeRef person = services.getPersonService().getPerson(entry.getKey());
                    services.getNodeService().setProperty(person, WorkspaceDiskUsage, entry.getValue());
                }

                for (Map.Entry<String, Long> entry : archive.entrySet()) {
                    NodeRef person = services.getPersonService().getPerson(entry.getKey());
                    services.getNodeService().setProperty(person, ArchiveDiskUsage, entry.getValue());
                }
                Calendar timeend = Calendar.getInstance();
                logger.error("Update diskusage has finished. Time taken: {} s", (timeend.getTimeInMillis() - timestart.getTimeInMillis()) / 1000);
                return null;
            }
        }, AuthenticationUtil.getSystemUserName());
    }

    @Uri(value = "/checkintegrity", method = HttpMethod.GET)
    public Resolution checkintegrity() {
        return new JsonWriterResolution() {
            @Override
            protected void writeJson(JSONWriter jsonWriter) throws JSONException {
                jsonWriter.array();
                for (MissingContent content : getIntegrityCheckResults()) {
                    jsonWriter.object();
                    jsonWriter.key("noderef").value(content.getNodeRef());
                    jsonWriter.key("contentUrl").value(content.getContentUrl());
                    jsonWriter.key("cause").value(content.getCause());
                    jsonWriter.endObject();
                }
                jsonWriter.endArray();
            }
        };
    }

    public List<MissingContent> getIntegrityCheckResults(){
        final ArrayList<MissingContent> missingContent = new ArrayList<>();
        List<Long> ids = jdbcTemplate.queryForList("select id from alf_node", Long.class);
        for (Long id : ids) {
            NodeRef nodeRef = nodeService.getNodeRef(id);
            if (nodeRef != null) {
                ContentDataWithId content = (ContentDataWithId) services.getNodeService().getProperty(nodeRef, ContentModel.PROP_CONTENT);
                if (content != null) {
                    try {
                        contentService.getReader(nodeRef, ContentModel.PROP_CONTENT).getContentInputStream().close();
                    } catch (Exception ex) {
                        String contentUrl = "<none>";
                        if (content != null && content.getContentUrl() != null) {
                            contentUrl = content.getContentUrl();
                        }
                        missingContent.add(new MissingContent(nodeRef, contentUrl, ex.getMessage()));
                    }
                }
            }
        }
        return missingContent;
    }
}

class MissingContent {
    private NodeRef nodeRef;
    private String contentUrl;
    private String cause;

    public MissingContent(NodeRef nodeRef, String contentUrl, String cause) {
        this.nodeRef = nodeRef;
        this.contentUrl = contentUrl;
        this.cause = cause;
    }

    public NodeRef getNodeRef() {
        return nodeRef;
    }

    public void setNodeRef(NodeRef nodeRef) {
        this.nodeRef = nodeRef;
    }

    public String getContentUrl() {
        return contentUrl;
    }

    public void setContentUrl(String contentUrl) {
        this.contentUrl = contentUrl;
    }

    public String getCause() {
        return cause;
    }

    public void setCause(String cause) {
        this.cause = cause;
    }
}

File: src/main/kotlin/eu/xenit/care4alf/module/ContentStore.kt
package eu.xenit.care4alf.module

import com.github.dynamicextensionsalfresco.annotations.AlfrescoService
import com.github.dynamicextensionsalfresco.annotations.ServiceType
import com.github.dynamicextensionsalfresco.webscripts.annotations.*
import com.github.dynamicextensionsalfresco.webscripts.resolutions.Resolution
import eu.xenit.care4alf.json
import eu.xenit.care4alf.web.LogHelper
import org.alfresco.model.ContentModel
import org.alfresco.query.PagingRequest
import org.alfresco.repo.domain.node.ContentDataWithId
import org.alfresco.service.cmr.notification.NotificationService
import org.alfresco.service.cmr.repository.ContentService
import org.alfresco.service.cmr.repository.NodeRef
import org.alfresco.service.cmr.repository.NodeService
import org.alfresco.service.cmr.security.PersonService
import org.alfresco.service.namespace.QName
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.ApplicationContext
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.stereotype.Component
import java.io.Serializable
import java.util.*
import javax.sql.DataSource

/**
 * @author Laurent Van der Linden
 */
@Component
@WebScript(baseUri = "/xenit/care4alf/contentstore", families = arrayOf("care4alf"), description = "Content store verification")
@Authentication(AuthenticationType.ADMIN)
public class ContentStore @Autowired constructor(
                                      private val applicationContext: ApplicationContext,
                                      dataSource: DataSource,
                                      private val personService: PersonService,
                                      private val notification: NotificationService,
                                      private val contentService: ContentService,
                                      @AlfrescoService(ServiceType.LOW_LEVEL) private val nodeService: NodeService
                            ) : LogHelper() {

    override val logger: Logger = LoggerFactory.getLogger(javaClass)

    private val jdbcTemplate = JdbcTemplate(dataSource)

    private val WorkspaceDiskUsage = QName.createQName("WorkspaceDiskUsage")
    private val ArchiveDiskUsage = QName.createQName("ArchiveDiskUsage")

    @Uri(value = "/diskusagebyowner")
    @Transaction(readOnly = true)
    fun list() = json {
        val people = personService.getPeople(null, true, null, PagingRequest(Integer.MAX_VALUE, null)).getPage().map({it.getNodeRef()})
        iterable(people) { person ->
            val workspace = nodeService.getProperty(person, WorkspaceDiskUsage) as? Long
            val archive = nodeService.getProperty(person, ArchiveDiskUsage) as? Long
            if (workspace != null || archive != null) {
                obj {
                  entry("username", nodeService.getProperty(person, ContentModel.PROP_USERNAME))
                  entry("workspace", workspace)
                  entry("archive", archive)
                }
            }
        }
    }

    @Uri(value = "/updatediskusage", method = HttpMethod.PUT)
    fun update() {
        val workspace = HashMap<String, Long>()
        val archive = HashMap<String, Long>()
        jdbcTemplate.queryForList("select id from alf_node", Long::class.java).forEach { id ->
            val nodeRef = nodeService.getNodeRef(id as Long)
            if (nodeRef != null) {
                val content = nodeService.getProperty(nodeRef, ContentModel.PROP_CONTENT) as? ContentDataWithId
                val owner = nodeService.getProperty(nodeRef, ContentModel.PROP_OWNER) as? String
                if (content != null && owner != null) {
                    if (nodeRef.getStoreRef().getProtocol().equals("archive")) {
                        archive.put(owner, archive.getOrElse(owner, {0L}) + content.getSize())
                    } else {
                        workspace.put(owner, workspace.getOrElse(owner, {0L}) + content.getSize())
                    }
                }
            }
        }

        workspace.forEach { ou ->
            val person = personService.getPerson(ou.key)
            nodeService.setProperty(person, WorkspaceDiskUsage, ou.value as Serializable)
            nodeService.let {  }
        }

        archive.forEach { ou ->
            val person = personService.getPerson(ou.key)
            nodeService.setProperty(person, ArchiveDiskUsage, ou.value as Serializable)
        }
    }

    @Uri(value = "/checkintegrity", method = HttpMethod.GET)
    fun checkintegrity(): Resolution {
        val missingContent = arrayListOf<MissingContent>()
        jdbcTemplate.queryForList("select id from alf_node", Long::class.java).forEach { id ->
            val nodeRef = nodeService.getNodeRef(id as Long)
            if (nodeRef != null) {
                val content = nodeService.getProperty(nodeRef, ContentModel.PROP_CONTENT) as? ContentDataWithId
                if (content != null) {
                    try {
                        contentService.getReader(nodeRef, ContentModel.PROP_CONTENT).getContentInputStream().close()
                    } catch(ex: Exception) {
                        var contentUrl="<none>";
                        if(content !=null && content.getContentUrl() != null)
                            contentUrl=content.getContentUrl();
                        missingContent.add(MissingContent(nodeRef, contentUrl, ex.message))
                    }
                }
            }
        }

        return json {
            iterable(missingContent) { missingContent ->
                obj {
                    entry("noderef", missingContent.noderef)
                    entry("contentUrl", missingContent.contentUrl)
                    entry("cause", missingContent.cause)
                }
            }
        }
    }
}

class MissingContent(val noderef: NodeRef, val contentUrl: String?, val cause: String?)
